### PR TITLE
Support setting device/fleet configuration in extra_uEnv.txt

### PR DESCRIPTION
**Pull request by 20k-ultra**
Closes: #1385
Change-Type: minor
Signed-off-by: Miguel Casqueira <miguel@balena.io>

### CODE REVIEW

**Comment by 20k-ultra** [comment]
I'd like to hear feedback on how I used [versionRequirements.ts](https://github.com/balena-io/balena-supervisor/blob/2dfd14c3d51c13f3e233a74ccf60e7fdc68f808d/src/lib/versionRequirements.ts) and why I made it. There are no places in the codebase that I know of that check versions to execute different code paths and I really dislike that it was introduced but not sure how else to approach the problem.

**Why:** I needed a way check if the device supported extra_uEnv in 2 places (extlinux, extra_uenv backends) so wanted them both to reference the same static value rather then us managing 2 static versions. These 2 places need to check if this feature is supported so they know if they should execute.

**How:** I created this centralized version requirement file that could be extended to include additional versions of other features we ant to check for.

Let me know your thoughts.

File: src/config/backends/extra-uEnv.ts [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+		const [, KEY, VALUE] = optionArray;
+		// Check if this key's value is a collection
+		if (ExtraUEnvConfigBackend.entries[KEY as EntryKey]?.collection) {
+			// Return split collection of options
+			return ExtraUEnvConfigBackend.parseOptionCollection(VALUE);
+		}
+		// Find the option that belongs to this entry
+		const optionKey = _.findKey(
+			ExtraUEnvConfigBackend.supportedConfigs,
+			(_v, k) => ExtraUEnvConfigBackend.supportedConfigs[k].key === KEY,
```
Comment (CameronDiver) [inline src/config/backends/extra-uEnv.ts:166]:
```suggestion
			(config) => config.key === KEY,
```

File: src/config/backends/extra-uEnv.ts [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+	}
+
+	private static configToString(configs: ConfigOptions): string {
+		// Get Map of ConfigOptions object
+		const configMap = ExtraUEnvConfigBackend.configToMap(configs);
+		// Iterator over configMap and concat to configString
+		let configString = '';
+		for (const [key, value] of configMap) {
+			if (!value) {
+				log.warn(`Could not find value for ${key} in configMap.`);
```
Comment (CameronDiver) [inline src/config/backends/extra-uEnv.ts:225]: This shouldn't be possible, right?

File: src/config/backends/extra-uEnv.ts [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+				// Check if we have to build the value for the entry
+				if (ENTRY_IS_COLLECTION) {
+					return configMap.set(
+						ENTRY_KEY,
+						// Append existing values if already set
+						(configMap.has(ENTRY_KEY) ? `${configMap.get(ENTRY_KEY)} ` : '') +
+							// Add new config's key
+							configKey +
+							// Append the config's value if it has one
+							(configValue !== '' ? `=${configValue}` : ''),
```
Comment (CameronDiver) [inline src/config/backends/extra-uEnv.ts:251]: There's a little bit too much going on here for my liking, I think even just another variable with a descriptive name would clear this up a little

File: src/config/backends/raspberry-pi.ts [hunk 4]
```diff
@@ -50,7 +50,10 @@ export class RPiConfigBackend extends DeviceConfigBackend {
 		'avoid_safe_mode',
 	];
 
-	public matches(deviceType: string): boolean {
+	public matches(
+		deviceType: string,
+		_metaRelease: string | undefined,
+	): boolean {
```
Comment (CameronDiver) [inline src/config/backends/raspberry-pi.ts:56]:
```suggestion
	public matches(deviceType: string): boolean {
```

File: src/lib/os-release.ts [hunk 5]
```diff
@@ -59,6 +59,12 @@ export function getOSSemver(path: string): Promise<string | undefined> {
 	return getOSReleaseField(path, 'VERSION');
 }
 
+export async function getMetaRelease(
```
Comment (CameronDiver) [inline src/lib/os-release.ts:62]:
```suggestion
export async function getMetaOSRelease(
```

File: src/lib/versionRequirements.ts [hunk 6]
```diff
@@ -0,0 +1,7 @@
+const VERSIONS = {
+	EXTRA_SUPPORT: {
```
Comment (CameronDiver) [inline src/lib/versionRequirements.ts:2]: This name isn't really descriptive enough for me. Looking at this file I (and without context) I don't really have any idea what this means, other than *something* to do with the jetson at v2.47

File: src/config/backends/extra-uEnv.ts (hunk 1 above)
Comment (20k-ultra) [reply src/config/backends/extra-uEnv.ts:166]: ahh... Why did I write that... thank you for catching that.

File: src/config/backends/extra-uEnv.ts (hunk 2 above)
Comment (20k-ultra) [reply src/config/backends/extra-uEnv.ts:225]: Very unlikely unless something had messed with the file. I was being extra verbose with this feature as I think more logs explaining why stuff doesn't happen is good.

File: src/config/backends/extra-uEnv.ts (hunk 3 above)
Comment (20k-ultra) [reply src/config/backends/extra-uEnv.ts]: Refactored it into another function.

File: src/lib/versionRequirements.ts (hunk 6 above)
Comment (20k-ultra) [reply src/lib/versionRequirements.ts]: I've removed this file and instead added a single value to constants.ts

https://github.com/balena-io/balena-supervisor/blob/cbe4fc4e5b276e9965402f105069d43ab2d3fc11/src/lib/constants.ts#L68-L69

**Comment by acostach** [comment]
In the current form, this PR works fine with a 2.52.7+rev1 build (custom dtb & isolcpus) and a 2.45.1+rev3 dashboard image (isolcpus).

**Comment by 20k-ultra** [comment]
> In the current form, this PR works fine with a 2.52.7+rev1 build (custom dtb & isolcpus) and a 2.45.1+rev3 dashboard image (isolcpus).

I would have expected this branch to still apply FDT for 2.45.1+rev3. If this is the case then I'll have to investigate this.

File: src/config/backends/backend.ts [hunk 7]
```diff
@@ -52,7 +52,7 @@ export abstract class DeviceConfigBackend {
 	): string | string[];
 
 	// Return the env var name for this config option
```
Comment (CameronDiver) [inline src/config/backends/backend.ts:54]: It'd be good if you could add a comment here that says what returning `null` actually means.

**Comment by 20k-ultra** [comment]
@balena-ci I self-certify!

**Pull request merged**